cpp/src/plasma/client.h

```diff
@@ -346,76 +309,19 @@ class ARROW_EXPORT PlasmaClient {
-  Status UnmapObject(const ObjectID& object_id);
-
   /// This is a helper method that flushes all pending release calls to the
   /// store.
   Status FlushReleaseHistory();
 
-  Status PerformRelease(const ObjectID& object_id);
-
-  /// Common helper for Get() variants
-  Status GetBuffers(const ObjectID* object_ids, int64_t num_objects, int64_t timeout_ms,
-                    const std::function<std::shared_ptr<Buffer>(
-                        const ObjectID&, const std::shared_ptr<Buffer>&)>& wrap_buffer,
-                    ObjectBuffer* object_buffers);
-
   bool IsInUse(const ObjectID& object_id);
 
-  uint8_t* lookup_or_mmap(int fd, int store_fd_val, int64_t map_size);
-
-  uint8_t* lookup_mmapped_file(int store_fd_val);
-
-  void increment_object_count(const ObjectID& object_id, PlasmaObject* object,
-                              bool is_sealed);
-
-  bool compute_object_hash_parallel(XXH64_state_t* hash_state, const unsigned char* data,
-                                    int64_t nbytes);
-
-  uint64_t compute_object_hash(const ObjectBuffer& obj_buffer);
-
-  /// File descriptor of the Unix domain socket that connects to the store.
-  int store_conn_;
-  /// File descriptor of the Unix domain socket that connects to the manager.
-  int manager_conn_;
-  /// Table of dlmalloc buffer files that have been memory mapped so far. This
-  /// is a hash table mapping a file descriptor to a struct containing the
-  /// address of the corresponding memory-mapped file.
-  std::unordered_map<int, ClientMmapTableEntry> mmap_table_;
-  /// A hash table of the object IDs that are currently being used by this
-  /// client.
-  std::unordered_map<ObjectID, std::unique_ptr<ObjectInUseEntry>, UniqueIDHasher>
-      objects_in_use_;
-  /// Object IDs of the last few release calls. This is a deque and
-  /// is used to delay releasing objects to see if they can be reused by
-  /// subsequent tasks so we do not unneccessarily invalidate cpu caches.
-  /// TODO(pcm): replace this with a proper lru cache using the size of the L3
-  /// cache.
-  std::deque<ObjectID> release_history_;
-  /// The number of bytes in the combined objects that are held in the release
-  /// history doubly-linked list. If this is too large then the client starts
-  /// releasing objects.
-  int64_t in_use_object_bytes_;
-  /// Configuration options for the plasma client.
-  PlasmaClientConfig config_;
-  /// The amount of memory available to the Plasma store. The client needs this
-  /// information to make sure that it does not delay in releasing so much
-  /// memory that the store is unable to evict enough objects to free up space.
-  int64_t store_capacity_;
-  /// Threadpool for parallel memcopy and hash computation.
-  std::vector<std::thread> threadpool_;
-#ifdef PLASMA_GPU
-  /// Cuda Device Manager.
-  arrow::gpu::CudaDeviceManager* manager_;
-#endif
+  class ARROW_NO_EXPORT Impl;
+  std::shared_ptr<Impl> impl_;
```

Comment:
Shouldn't PIMPLs be at least a `unique_ptr`?

Note that we might should use something of the lines of http://brotherst-code.blogspot.de/2015/09/the-pimpl-idiom-youre-doing-it-wrong.html as we would still run into the problem of const correctness with PIMPLs otherwise.

Reply:
Yeah, we will need the shared_ptr so I'm in favor of leaving it. I wonder if there is a good way to get the PIIMPL const correctness with the shared pointer.

Reply:
For `PlasmaClient` I don't think const-ness is a very useful piece of information. It's a thing that talks over the network (what is a const TCP connection?).

Reply:
Yeah, I agree!

Reply:
Yes, this was a more general piece of information. The const correctness is rather something that we would need with the core Arrow structures such as Array or Buffer.